### DISCUSSION TITLE
Possibility of writing human readable grammar rules from deinflect.ts

### DISCUSSION THREAD

**Discussion by Kimeiga** [birchill/10ten-ja-reader#2048] [Ideas]
Hi guys I just want to start by saying thank you so dearly for this project. Actually all of the chinese popup readers that I know of have been killed by the move to manifest v3 by chrome and I noticed right away that 10ten reader is still working, not sure how you guys pulled off this magic.

I am trying to study Japanese verb and adjective conjugations and noticed a lot of different ways to study them as indicated online. There's sort of a traditional approach of learning the different verb bases (Mizenkei, Kateikei, etc) (https://en.wikipedia.org/wiki/Japanese_conjugation) and then learning which auxiliary verbs can attach to which bases. 

There also seems to be a way to understand it using "modern linguistic analysis" where you are willing to break up the mora in the case of godan verbs and imagine the stem of [書く](https://en.wiktionary.org/wiki/%E6%9B%B8%E3%81%8F#Japanese) (kaku) as kak-, and the "conjugations" would then include the vowel. 
https://en.wiktionary.org/wiki/Appendix:Japanese_verbs

I've watched a lot of youtube video teachers as well and there's some dissonance about which way to teach it. 

However when using 10ten reader I feel like I discover new conjugations I didnt know previously. For instance yesterday while using 10ten reader I discovered the conjugation: 待たされる which is 待たせられる where せら〜 contracts into さ〜, where I hadn't been taught that explicitly anywhere online (albeit I did find it on wikipedia later).

Given you guys have done so much work with deinflect.ts and word-search.ts to figure out japanese verb conjugations completely,

**I was wondering if it made any sense for one to try to convert the knowledge contained in those files into some kind of human readable understandable learning tool for how to conjugate verbs?**

Or would it be too difficult to create something human understandable out of it and it's better to learn it a different way? If this is true, is it because the rules are somewhat over-redundant for the purpose of ensuring correctness, as the deinflect.ts file is quite long? For instance, I see 
`['ませんでした', '', Type.Initial, Type.MasuStem, [Reason.PolitePastNegative]],`
`['ません', '', Type.Initial, Type.MasuStem, [Reason.PoliteNegative]],`
mentioned separately. I suppose as a learner you might not learn these as separate rules but rather learn ません and learn です's past tense and put them together. 

Any thoughts on this? And if it wouldn't make much sense, do you guys have any grammar resources that you would think are the best to learn the rules from? Again thank you so much!!

**Comment by birtles** [comment]
Hi Kimeiga! Thank you so much!

I think @enellis would have a much better idea about the feasibility of repurposing those rules. My impression is that unfortunately it might be difficult for the reasons you mentioned.

**Comment by enellis** [comment]
Hey Kimeiga! First of all, I have to agree with you that @birtles really is some kind of magician, who puts so much work into the little details, listening patiently to every feedback he gets, so this tool works out of the box for everyone without any problems. Thanks to him for maintaining this project in such a friendly and attentive manner!

I am not really sure what is meant by "conjugation explainer" in #1970, but I assume this idea goes in the same direction. As a first step of working towards something like this, I will summarize some of my thoughts about Japanese inflections.

I was working on the `deinflect.ts` file in the last few months and actually have a very strong opinion on what the easiest way to understand Japanese inflections is. The thing is, it is actually super easy and very logical. Only because of that we are able to parse a big part of all Japanese inflections in a ruleset as small as about 240 rules. And most of them are just special irregular cases caused by some archaic linguistical structures surviving into modern times. (Well, that is the reason for many irregularities in any language I suppose.) And most of these are actually exclusively used in honorific speech, so you don't even have to worry about them as a beginner.

In my opinion, the biggest mistake you can make, is trying to learn Japanese inflections as you would do with western languages through conjugation tables. Japanese verbs don't really conjugate. There are the stems and the auxiliary verbs that attach to them. It's a very regular and straightforward system, far more simple than the irregular conjugation mess you have to deal with when learning any western language. Don't learn Japanese from resources that are "westernizing" the language!

I think, the best approach of learning the language is in a way Japanese people themselves learn about their language. Of course, in the beginning, you are not able to approach Japanese resources, but you should try to get there as soon as possible. To do that, you should learn the following simplified version of the grammatical structure of Japanese, which helps you becoming being able to "parse" and break down Japanese text in a very short amount of time. As long as you have a firm grasp of the grammatical structure, you will be able to understand _any_ Japanese text with the help of 10ten quite quickly.

### 五十音 - The heart of Japanese
_(Read top-to-bottom / right-to-left. あ, い, う, え, お, か, き, く, け, こ, ...)_
|   | **W** | **R** | **Y** | **M** | **P** | **B** | **H** | **N** | **D** | **T** | **Z** | **S** | **G** | **K** |   |   |
| :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: |
| ん | わ | ら | や | ま | ぱ | ば | は | な | だ | た | ざ | さ | が | か | あ | **a** |
|   |   | り |   | み | ぴ | び | ひ | に | ぢ | ち | じ | し | ぎ | き | い | **i** |
|   |   | る | ゆ | む | ぷ | ぶ | ふ | ぬ | づ | つ | ず | す | ぐ | く | う | **u** |
|   |   | れ |   | め | ぺ | べ | へ | ね | で | て | ぜ | せ | げ | け | え | **e** |
|   | を | ろ | よ | も | ぽ | ぼ | ほ | の | ど | と | ぞ | そ | ご | こ | お | **o** |

Learning Hiragana and Katakana should _always_ be the very first step when learning Japanese. Knowing Hiragana and the system behind it _by heart_ is essential when trying to get behind the logical structure of Japanese grammar. Don't learn Japanese using Rōmaji, you will become confused!

Everyone learning Japanese should already know this, so this is giving just a very brief overview.

Most Japanese syllables, or, to be more precise, a _mora_ starts with a consonant and ends with a vowel. The rows of the table represent the 5 vowels of the Japanese language and the columns represent the consonants. The exceptions are the first column's あ, い, う, え, お, which are just the single vowel sounds without a preceding consonant, and the last letter ん, which is a little special in that it cannot be at the beginning of a word and that it has several different pronounciations depending on what follows. ん doesn't really belong into the `a` column, there are no relations with the "a" vowel sound.

Excluding Yōon moras (e.g. しゃ, きゅ, りょ), which are not really of interest when looking at inflection, **this table represents all sounds of the Japanese language**.

### What word types to worry about?
```mermaid
%%{init: {'themeVariables': { 'fontSize': '12px', 'fontFamily': 'Inter'}}}%%
graph TD;
    di[No inflection] --> id1["Nouns / 名詞"];
    Inflection --> id2["Verbs / 動詞"];
    Inflection --> id3["Adjectives / 形容詞"];
    Inflection --> id4["Copula / だ"];
```
As a start it is best to categorize Japanese words into 4 basic categories. Nouns, verbs, adjectives and the copula だ. We don't really have to worry about nouns. They don't inflect. Verbs, adjectives and the copula do. Let's have a closer look at verbs first.

### Verbs
```mermaid
%%{init: {'themeVariables': { 'fontSize': '12px', 'fontFamily': 'Inter'}}}%%
graph TD;
    id5["Verbs / 動詞"];
    id5 --> id2["Godan / 五段"];
    id5 --> id1["Ichidan / 一段"];
    id5 --> id3["suru / する"];
    id5 --> id4["kuru / くる"];
```
First, a few premises:
- Every plain dictionary form of a Japanese verb ends in a kana of the `u` row of the Japanese syllabary.
- Every verb is either of type Ichidan or Godan.
_(Except the verbs `する - to be` and `くる - to come`. These are together with the polite helper verb `ます` one of the very few commonly used verbs with irregular inflections.)_
- Every verb posesses different stems depending on the type of the verb.
- Auxiliaries, indicating the mood of the verb, attach to a specific stem type.

#### Godan verbs
Examples: `買う - to buy`, `書く - to write`, `泳ぐ - to swim`, `話す - to speak`, `勝つ - to win`, `死ぬ - to die`, `選ぶ - to choose`, `読む - to read`, `変わる - to change`

The stems of Godan verbs can easily be determined by looking at the specific column of the ending kana of the verb in dictionary form. Remember how there are 5 rows in the Japanese syllabary. These are the five stems!
Let's take all the examples from above. Have a look at their respective columns in the Japanese syllabary:
|        | 買う | 書く | 泳ぐ | 話す | 勝つ | 死ぬ | 選ぶ | 読む | 変わる |
|  :---  |:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|
| a-stem | _買わ_ | 書か | 泳が | 話さ | 勝た | 死な | 選ば | 読ま | 変わら |
| i-stem | 買い | 書き | 泳ぎ | 話し | 勝ち | 死に | 選び | 読み | 変わり |
| u-stem | 買う | 書く | 泳ぐ | 話す | 勝つ | 死ぬ | 選ぶ | 読む | 変わる |
| e-stem | 買え | 書け | 泳げ | 話せ | 勝て | 死ね | 選べ | 読め | 変われ |
| o-stem | 買お | 書こ | 泳ご | 話そ | 勝と | 死の | 選ぼ | 読も | 変わろ |

_One small exception:_ The a-stem of verbs from the first column of the Japanese syllabary (single vowel moras) is not あ but rather わ for a better pronunciation. That means that the a-stem of 買う is **not** 買あ but rather 買わ.

We end up with 5 stem types -> 五段.

#### The te/past-stem
There are two essential verb forms that cannot be constructed using the previously mentioned five stems: the past tense and the te-form. These forms share a specialized stem, which varies depending on the verb's final kana. The final kanas of Godan verbs can be categorized into "hard" sounds (う, つ, る, し, く) and "soft" sounds (ぐ, ぬ, ぶ, む). For verbs with hard ending sounds, the te-form and past tense are created by adding て and た to this specialized stem. For soft sounds, the suffixes て and た become softened to で and だ.

To construct the stem, replace the final
- う, つ, る with っ.
  - te-form:　買って、勝って、変わって
  - past:　　  買った、勝った、変わった
- す with し.
  - te-form:　話して
  - past:　　  話した
- く (hard) with い.
  - te-form:　書いて
  - past:　　  書いた
  - _Exception: `行く - to go`. く changes into っ instead of い._
    - _te-form:  行って_
    - _past:　　行った_
- ぐ (soft) with い.
  - te-form:　泳いで
  - past:　　  泳いだ
- ぬ, ぶ, む with ん.
  - te-form:　死んで、選んで、読んで
  - past:　　  死んだ、選んだ、読んだ


#### Ichidan verbs
Examples: `見る - to see`, `着る - to wear`, `食べる - to eat`, `考える - to think`

Every Ichidan verb ends in -iru or -eru, but not every verb ending in -iru or -eru is an Ichidan verb. `走る - to run` for example is a Godan verb. In these cases there is no other way than to explicitly remember the word type.

The stems of Ichidan verbs are super simple. They are all the same, i.e. there is only one: the dictionary form without the final る.

Let's again look at all the examples above:
|   | 見る | 着る | 食べる | 考える |
| :--- |:---:|:---:|:---:|:---:|
| a / i / u / e / o / te/past-stem | 見 | 着 | 食べ | 考え |

We end up with 1 stem type -> 一段.

_Be prepared though!_ Ichidan verbs utilize different auxiliaries in some cases.

#### Auxiliaries

This is an overview of what auxiliary word attaches to which stem type.

- Dictionary form
  - Negative imperative: な
- a-stem
  - Negation: ない (adjective), ぬ, ず (noun-like)
  - Passive:
    - Godan: れる (Ichidan verb)
    - Ichidan: られる (Ichidan verb) - **Same as the potential form**
  - Causative:
    - Godan: せる (Ichidan verb)
    - Ichidan: させる (Ichidan verb)
  - Godan: Shortened form of せられる (Causative + Passive): される (Ichidan verb)
- i-stem
  - Expressing politeness: ます (irregular)
  - Expressing desire: たい (adjective)
  - Expressing excess: すぎる (Ichidan verb)
  - Expressing appearance / conjecture: そう (noun-like)
- e-stem
  - Imperative:
    - Godan: plain stem
    - Ichidan: ろ, よ
  - Potential:
    - Godan: る (Ichidan verb)
    - Ichidan: られる (Ichidan verb) - **Same as the passive form**
      - ら抜き言葉: In casual speech, the ら in the potential form is sometimes omitted, so られる is shortened to れる.
  - Provisional:
    - Godan: ば
    - Ichiban: れば
- o-stem:
  - Volitional:
    - Godan: う
    - Ichidan: よう
- te/past-stem:
  - Te-form: て / で
    - Continuous (te-form + 居る): ている / でいる (Ichidan verb)
      - In spoken Japanese, the い is often omitted, so it is shortened to てる / でる.
  - Past: た / だ
  - Conditional: たら / だら
  - Examplification or expressing repetition: たり / だり

#### Irregularities of 来る
The verb `くる - to come` is inflected slightly differently from other verbs. The easiest way to remember its inflections is to think of it as an Ichidan verb with changing stems.

| Dictionary form | a / o-stem | i / te/past-stem | e-stem | Provisional | Imperative |
| :---: | :---: | :---: | :---: | :---: | :---: |
| くる | こ | き | ❌ | くれば | こい |

Examples: こない, こられる, こさせる, こよう, きます, きて, きたら, くれば, こい.

#### Irregularities of 為る
We can think of `する - to do` as an Ichidan verb as well, but it has more irregularities, particularly in the a-stem. Instead of attaching られる and させる, we use れる and せる. Additionally, a distinct potential form for する is used: できる (Ichidan verb).

| Dictionary form | a-stem | i / o / te/past-stem | e-stem | Potential | Provisional | Imperative |
| :---: | :---: | :---: | :---: | :---: | :---: | :---: |
| する | し（-ない）、さ（-れる、-せる）、せ（-ぬ、-ず） | し | ❌ | できる | すれば | しろ, しよ |

Examples: しない, される, させる, しよう, します, して, したら, すれば, しろ.

#### Irregularities of ます
The auxiliary verb `ます` is used mainly at the end of a sentence to express politeness toward the listener and is attached to the i-stem of a verb. Since it primarily serves to add a polite tone to the verb, there are relatively few forms of ます in use. The archaic negative form makes this special inflection pattern particularly unique.

| Dictionary form | te/past-stem | Negative | Negative past | Volitional | Provisional | Imperative |
| :---: | :---: | :---: | :---: | :---: | :---: | :---: |
| ます | まし（-て、-た、-たら） | ません | ませんでした | ましょう | ますれば | ませ, まし |

Examples: 見**ました** - did see, 食べ**ません** - does not eat, 読み**ませんでした** - did not read, 行き**ましょう** - let's go, いらっしゃい**ませ** - Come! (used as a welcome in shops, restaurants, etc.)

### Adjectives
Examples: `嬉しい - happy`, `新しい - new`, `可愛い - cute`, `高い - high`, `長い - long`, `難しい - difficult`, `忙しい - busy`, `早い - fast`, `強い - strong`

There are two types of adjectives in Japanese: i-adjectives (形容詞) and na-adjectives (形容動詞). Na-adjectives are fundamentally nouns and do not inflect on their own. To use them, a special form of the copula だ is attached, which governs their inflection. The copula, however, is covered in a different section. This section focuses on i-adjectives.

The dictionary form of i-adjectives ends with the kana い. By removing the い, you get the stem, to which you can attach the various forms shown in the table below.

| Dictionary form | Adverbial | te-form | Negative | Past / Conditional / Examplification | Conjectural | Provisional |
| :---: | :---: | :---: | :---: | :---: | :---: | :---: |
| い | く | くて | く（-ない） | かっ（-た、-たら、-たり） | かろう | ければ |

Examples: 嬉しく - happily, 新しくない - not new, 可愛かった - was cute

Other auxiliaries, such as すぎる and そう, attach directly to the stem. Examples: 高すぎる - too high, 難しそう - seems difficult.

### Playing Lego
This is the most fun part of Japanese: You can play Lego with it – auxiliaries attach other auxiliaries.

Let's say for example, you want to negate the adjective 嬉しい in the past tense. The first step is to build the negative form: 嬉し~~い~~ &rarr; 嬉し**くない**. ない itself is again an adjective, which means you can build the past form like this: 嬉しくな~~い~~ &rarr; 嬉しくな**かった**.

As another example, how would you politely express that you were drinking too much last night? First, remember the word type of `飲む - to drink` – a Godan verb. You want to add すぎる, the auxiliary verb expressing excess, which attaches to the i-stem: 飲~~む~~ &rarr; 飲**み** &rarr; 飲み**すぎる**. すぎる itself is an Ichidan verb. Attach the polite helper verb ます to it: 飲みすぎ~~る~~ &rarr; 飲みすぎ &rarr; 飲みすぎ**ます**. After forming the past tense of ます, you are done: 飲みすぎま~~す~~ &rarr; 飲みすぎま**した**.

**This is very much work in progress. To be continued and updated...**

---
My understanding of Japanese is _heavily_ based on [Cure Dolly's Course On Organic Japanese](https://www.youtube.com/watch?v=pSvH9vH60Ig&list=PLg9uYxuZf8x_A-vcqqyOFZu06WlhnypWj). Even though the videos are not really that accessible, as the teacher takes the form of a V-Tuber and the audio quality can be quite rough and hard to understand sometimes, I think Cure Dolly's knowledge of Japanese was quite profound and admireable. Regrettably, she died in 2021 because of an illness. Her understanding itself was based on Dr. Jay Rubin's _Making Sense of Japanese: What the Textbooks Don't Tell You_ (1998).

**Reply by StarScape** [reply]
I'm not involved here and have nothing to add, but I just want to say this amazing and detailed response from one complete stranger on the internet to another is what makes me love open source ❤️.

**Reply by Kimeiga** [reply]
enellis you are amazing man I will 100% scour through this in a bit. I've also watched a lot of cure dolly and read through the wikipedia article on [japanese verb bases](https://en.wikipedia.org/wiki/Japanese_conjugation) that introduced me to what you're talking about. 

I had some questions about this though:

if れる is an ichidan verb, and it could also appear as られる, but the causative form of する is られる does that mean that this auxiliary verb is really just the causative する, and for some reason we drop ら with 未然形 form of godan verbs? or is れる truly a distinct verb from する's causative form?

And this question could be extended to (さ)せる. and also (られ)る haha for the potential form that goes on 命令形.

And then a second question is kinda like:

Does it make more sense to use the "base" system and auxiliary verbs to understand japanese inflection, or if we break the mora into consonant and vowel would it "make more sense"?

I'm Turkish and we have an agglutinating language as well (albeit even more agglutinative than Japanese). Our particles which we inflect verbs with, like Japanese, often have "spacing" letters that you would use depending on the ending of the word you're attaching it to. 

For instance, -(y)ecek means future tense, and we add the "y" if the verb ends in a vowel to be pronounced between the last vowel of the verb and the first of this particle "ecek"

Ne diyeceksin?
What are you going to say?

Dictionary form of the verb: demek (to say)

Stem of the verb: de-
However, the root "de-" changes to "di-" in certain conjugations due to vowel harmony (common for irregular verbs)

Future tense suffix: -(y)ecek
Since the stem "de-" ends with a vowel, we add the buffer letter "y" before the future tense suffix -ecek, becoming diyecek.

Personal suffix: -sin (for second-person singular "you")
In the future tense, the second-person singular ending is -sin.

Ben ölecek miyim?
Am I going to die?

Dictionary form of the verb: ölmek (to die)

Stem of the verb: öl-

Future tense suffix: -(y)ecek
The stem "öl-" does not end with a vowel, so no buffer letter "y" is needed. It simply becomes ölecek.

Personal suffix: -im (for first-person singular "I")
In the future tense, the first-person singular ending is -im.

As a Turkish person I feel like there's something similar happening in Japanese with the difference between the godan and the ichidan auxiliary verbs

I understand this is not at all how japanese people learn their language but hear me out

<table class="wikitable">
<tbody><tr>
<th>Conjugational class</th>
<th>Dictionary form</th>
<th>Stem</th>
<th>Conjugated forms
</th></tr>
<tr>
<td>Consonant-stem</td>
<td><i class="Jpan mention" lang="ja"><a href="/wiki/%E6%9B%B8%E3%81%8F#Japanese" title="書く">書く</a></i> <span class="mention-gloss-paren annotation-paren">(</span><span class="mention-tr tr">kaku</span><span class="mention-gloss-paren annotation-paren">)</span></td>
<td><i>kak-</i></td>
<td><i>kakanai<br>kakimasu<br>kaku<br>kakeba<br>kakō</i>
</td></tr>
<tr>
<td>Vowel-stem</td>
<td><i class="Jpan mention" lang="ja"><a href="/wiki/%E8%B5%B7%E3%81%8D%E3%82%8B#Japanese" title="起きる">起きる</a></i> <span class="mention-gloss-paren annotation-paren">(</span><span class="mention-tr tr">okiru</span><span class="mention-gloss-paren annotation-paren">)</span></td>
<td><i>oki-</i></td>
<td><i>okinai<br>okimasu<br>okiru<br>okireba<br>okiyō</i>
</td></tr></tbody></table>

just grabbing the [table from wiktionary here ](https://en.wiktionary.org/wiki/Appendix:Japanese_verbs)

it does appear to me that you could "understand" the auxiliary verbs in the "turkish lens" in the following way:

if you treat the "stem" of godan verbs as the part before "u" i.e. kaku -> "kak-", and for ichidan verbs the part before "ru" i.e. okiru -> "oki"

then can't you "merge" the verb base ending with the auxiliary verbs themselves (considering they can only attach to certain bases anyways).

verbs:
causative: -(a)ser(u)
passive/receptive: -(a)rer(u)
...

adjectives:
desire: -(i)ta(i)

where the "rule" would be that, for the initial vowel, you would only have it if otherwise it would be two consonants together (i.e. the "padding" concept of the "y" in turkish). 

And if there are two optional vowels the following one always takes precedence. And if the optional vowel is at the end it always displays. 

verb i.e.
mat- + (a)ser(u) + (a)rer(u) => mataserareru 

adj i.e.
tabe- + (i)ta(i) => tabetai

I understand that this isn't what's actually happening, but one thing about this system that the normal way requires you to understand what auxiliary verbs can attach to which verb stems, but this way (if it was sound logic) would make it so that you could combine the verb stem and the auxiliary verb into one mental concept

But one thing that doesn't really work so well is that technically せる is an ichidan verb, but in order to create the causative passive properly, we had to make the final "u" the parenthetical (-(a)ser(u)) instead of writing (-(a)se(ru)) since i just earlier said that ichidan verb's stems in this system are the part before "ru"

It was just that I noticed that it was mentioned [here](https://en.wiktionary.org/wiki/Appendix:Japanese_verbs) as an example of modern linguistic analysis differing from the traditional way of teaching japanese, and I was curious if this way of conceptualizing things could capture the full complexity of the inflection system in fewer rules than the traditional system like [as featured here](http://cghq.net/japanese/verbs/AMB_Japanese_Verbs.pdf)

would love to hear other's thoughts though! And again thank you for the very helpful response!

**Reply by enellis** [reply]
> if れる is an ichidan verb, and it could also appear as られる, but the causative form of する is られる does that mean that this auxiliary verb is really just the causative する, and for some reason we drop ら with 未然形 form of godan verbs? or is れる truly a distinct verb from する's causative form?
> 
> And this question could be extended to (さ)せる. and also (られ)る haha for the potential form that goes on 命令形.

れる, られる, せる and させる are all distinct verbs with no relation to する.

One of the 未然形s of する is さ. When we attach れる to the 未然形 of する we get される. The passive of する.
When we attach せる to the 未然形 of する we get させる. The causative of する. This is **not** the same verb as the auxiliary verb させる, which originated from the classical Japanese verb さす.

Interestingly, thanks to the rich history of the する-inflection pattern, there are 2 other 未然形 of する: せ and し. You can form other passives of する with them, using not れる but られる as the auxiliary: せられる and しられる.

> Does it make more sense to use the "base" system and auxiliary verbs to understand japanese inflection, or if we break the mora into consonant and vowel would it "make more sense"?

First of all, thank you very much for this very interesting dive into Turkish grammar. Living in Germany, I encounter Turkish being spoken in public quite frequently, but sadly, I've never really dug deeper and started learning about it. As a result, I don't think I could fully grasp the point you were making, as the concept of "spacing letters" is completely new to me, so I will have to give a more generalized answer.

The Japanese language is completely isolated with no relations to any other language. Every attempt forcing the inflection system of another language upon the Japanese's inflection system is probably going to make it seem more complicated than it actually is, with a lot of exceptions and irregularities.
The beauty of the Japanese inflections system is, that it is for the most part completely regular. This is only if the moras are not broken into consonants and vowels, written out in Latin letters.

I have a (maybe a little bit dumb) question. Let's take 話す for example. If you'd say "hanas-" is the stem, could you explain with your approach the connection to (i)ta(i): hanas-h-(i)ta(i)? What about the h? Wouldn't this be just another irregularity?

**Reply by birtles** [reply]
> Hey Kimeiga! First of all, I have to agree with you that @birtles really is some kind of magician, who puts so much work into the little details, listening patiently to every feedback he gets, so this tool works out of the box for everyone without any problems. Thanks to him for maintaining this project in such a friendly and attentive manner!

Thank you so much @enellis, you are much too kind! Thank you for all your incredible contributions—especially making our deinflection code so much more accurate and robust!

**Reply by enellis** [reply]
@birtles I'm wondering if it might be a good idea to start a wiki and explain the deinflection done by 10ten in a similar way to how I've tried here. We could also create individual articles for the various auxiliary words, providing detailed explanations of how they work, with the long-term goal of eventually integrating this information somehow into 10ten. What do you think? I can understand if you feel it might stretch the scope of the project a bit.

**Reply by birtles** [reply]
@enellis That sounds awesome. Stretching things a bit further, would there be some sort of human readable format we can synchronize with the code? Or perhaps the other way around—e.g. having markdown code comments we can extract along with the corresponding rules to generate a human readable summary of what is going on?

**Reply by enellis** [reply]
@birtles, That's a good idea—the explanations for the auxiliaries could be added as markdown comments for each member of the Reason enum. These comments could be parsed by a tool like [TypeDoc](https://typedoc.org/guides/overview/), allowing us to display the generated information when, for example, hovering over a reason in the reason chain.

I'm not a big fan of using submodules, but another option could be to parse the markdown pages from the Wiki Git repository, added as a submodule, using something like [markdown-it](https://github.com/markdown-it/markdown-it). A GitHub Action syncing a `/wiki` folder and the wiki git would do the trick to, I guess.

**Reply by Kimeiga** [reply]
>  Let's take 話す for example. If you'd say "hanas-" is the stem, could you explain with your approach the connection to (i)ta(i): hanas-h-(i)ta(i)? What about the h? 

i wrote a long answer to this and then closed the tab but basically in that case it would just be hanas + -(i)ta(i) = hanasitai where "si" is pronounced "shi" in modern japanese. The point of the latin letters is just to be able to bisect the mora and think of godan verbs as having a stem that ends in a consonant and ichidan verbs as having their stem ending in a vowel. Indeed it seems there is some evidence that linguists analyse japanese in this manner

https://www.reddit.com/r/LearnJapanese/comments/1csf05/comment/c9jq3gu/?utm_source=share&utm_medium=web3x&utm_name=web3xcss&utm_term=1&utm_content=share_button

https://books.google.com.br/books?id=SszxbMtHbs8C&lpg=PP1&pg=PA308&redir_esc=y#v=onepage&q&f=false

you can see that the linguist in the book linked is discussing transitive-intransitive verb pairs in japanese and uses this structure like here:

<img width="875" alt="image" src="https://github.com/user-attachments/assets/1a5ec9c9-0260-4bc2-9673-343541585014">

where he says -as- makes intransitive verbs transitive e.g. 乾く kawak- + -as = kawakas- 乾かす where 乾く is intransitive and 乾かす is transitive. Using this structure kinda merges the -a of the irrealis with the s of す with the knowledge that if another suffix was to be added to the end of this word, the す (or whatever kana is the final mora of the suffix) would have to be inflected to the correct verb base anyways to accept the next suffix. 

i dont really know a whole lot about linguistics here but i can see this as an interesting opportunity to cut down on the amount of things to remember (which verb stem accepts which suffix -> just memorize them as a pair put onto a stem). however I wouldn't know a whole lot about if this method of thinking would produce any incorrect conclusions

**Reply by enellis** [reply]
> i wrote a long answer to this and then closed the tab

Oh no, I feel you, so frustrating when this happens.


> i dont really know a whole lot about linguistics here but i can see this as an interesting opportunity to cut down on the amount of things to remember (which verb stem accepts which suffix -> just memorize them as a pair put onto a stem). however I wouldn't know a whole lot about if this method of thinking would produce any incorrect conclusions

It's an interesting perspective, and in the end, there really is no right or wrong. If you are able to remember and comprehend Japanese grammar analyzed this way best, then you should learn Japanese from resources that do so.

What I'm not seeing, though, and this is the most important part when it comes to 10ten, is how that way of thinking about inflections could reduce the ruleset further than the stem system. Isn't remembering, for example, that たい attaches to the i-stem and せる attaches to the a-stem the same amount of information (especially from a programmer's standpoint) as remembering (i)tai and (a)seru?

I'm sorry for being so opinionated toward the traditional stem system; it's just that it has proven very useful (we were able to reduce the deinflection ruleset by ~60% after integrating abstractions for the te/past-stem (連用形の音便), i-stem (連用形), and a-stem (未然形) for the first time). It could very well be that I just haven't fully grasped the benefits of that kind of analysis of Japanese yet.
